rust/src/dataset/optimize.rs

```diff
@@ -0,0 +1,1015 @@
+        let mut bins = Vec::new();
+
+        loop {
+            let mut bin_len = 0;
+            let mut bin_row_count = 0;
+            while bin_row_count < min_num_rows && bin_len < self.row_counts.len() {
+                bin_row_count += self.row_counts[bin_len];
+                bin_len += 1;
+            }
+
+            // If there's enough remaining to make another worthwhile bin, then
+            // push what we have as a bin.
+            if self.row_counts[bin_len..].iter().sum::<usize>() >= min_num_rows {
+                bins.push(Self {
+                    fragments: self.fragments.drain(0..bin_len).collect(),
+                    pos_range: self.pos_range.start..(self.pos_range.start + bin_len),
+                    candidacy: self.candidacy.drain(0..bin_len).collect(),
+                    row_counts: self.row_counts.drain(0..bin_len).collect(),
+                });
+                self.pos_range.start += bin_len;
+            } else {
+                // Otherwise, just push the remaining fragments into the last bin
+                bins.push(self);
+                break;
+            }
+        }
+
+        bins
+    }
+}
+
+/// Formulate a plan to compact the files in a dataset
+///
+/// The compaction plan will contain a list of tasks to execute. Each task
+/// will contain approximately `target_rows_per_fragment` rows and will be
+/// rewriting fragments that are adjacent in the dataset's fragment list. Some
+/// tasks may contain a single fragment when that fragment has deletions that
+/// are being materialized and doesn't have any neighbors that need to be
+/// compacted.
+pub async fn plan_compaction(
+    dataset: &Dataset,
+    options: &CompactionOptions,
+) -> Result<CompactionPlan> {
+    // We assume here that get_fragments is returning the fragments in a
+    // meaningful order that we want to preserve.
+    let mut fragment_metrics = futures::stream::iter(dataset.get_fragments())
+        .map(|fragment| async move {
+            match collect_metrics(&fragment).await {
+                Ok(metrics) => Ok((fragment.metadata, metrics)),
+                Err(e) => Err(e),
+            }
+        })
+        .buffered(num_cpus::get() * 2);
+
+    let mut candidate_bins: Vec<CandidateBin> = Vec::new();
+    let mut current_bin: Option<CandidateBin> = None;
+    let mut i = 0;
+
+    while let Some(res) = fragment_metrics.next().await {
+        let (fragment, metrics) = res?;
+
+        let candidacy = if options.materialize_deletions
+            && metrics.deletion_percentage() > options.materialize_deletions_threshold
+        {
+            Some(CompactionCandidacy::CompactItself)
+        } else if metrics.fragment_length < options.target_rows_per_fragment {
+            // Only want to compact if their are neighbors to compact such that
+            // we can get a larger fragment.
+            Some(CompactionCandidacy::CompactWithNeighbors)
```

Comment:
Can a small fragment be compacted into a big fragment?  In other words, if we have:

Fragment 1: 2 million rows
Fragment 2: 1000 rows
Fragment 3: 2 million rows

And if the target is 1 million rows will it merge fragment 2 into fragment 3? Or does "neighbors" mean "neighbors that also need compacting?"

Reply:
Right now this won't compact any, which might not be ideal for that situation. `CompactWithNeighbors` means only run compaction if there is a neighbor that also needs it. So if neither neighbors are candidates, this will skip it. 

The usual pattern we care about is going to look like:

```
# fragments written before that were compacted
Fragment 1: 2 million rows
Fragment 2: 2 million rows
# fragments recently appended
Fragment 3: 1000 rows
Fragment 4: 1000 rows
Fragment 5: 1000 rows
```

So we expect the small fragments to be contiguous.

Reply:
Sounds good.  We might just want a note in some doc comment somewhere.  Some future power user is inevitably going to ask about it :)

Reply:
Added some notes about that.